### ISSUE TITLE
Analyser dosent work in files where it uses an self created class method named "catchError" 

### ISSUE THREAD

**Issue by henrikskul** [dart-lang/sdk#46051]
this issue was originally postet at https://github.com/flutter/flutter/issues/82443, but i was told to post i here instead

## Steps to Reproduce

1. create an class with an method named "catchError".
2. use this method in another file.
3. .The analyser/linter dosent work in this file, because it uses the self created catchError method.

**Expected results:** That the analyser should work. If "catchError" should not be a name of an self created method, the analyser should tell so.

**Actual results:** Analyser dosent detect errors in the file that uses the self created catchError method.

I have reproduced the error here https://github.com/henrikskul/Flutter_catch_error_bug.

#### Output of `flutter analyze` in the project

```
flutter analyze
Analyzing flutter_catc_error_bug...
No issues found! (ran in 3.1s)
```

#### Output of `flutter run` in the project

<details> 
<summary>logs</summary> 
	
```console
flutter run
Using hardware rendering with device Android SDK built for x86. If you notice graphics artifacts, consider enabling software rendering with "--enable-software-rendering".
Launching lib\main.dart on Android SDK built for x86 in debug mode...
Running Gradle task 'assembleDebug'...
lib/ButtonWithoutLinting.dart:15:11: Error: 'the' isn't a type.

          the linter should show an error here

          ^^^

lib/ButtonWithoutLinting.dart:15:15: Error: Expected ';' after this.

          the linter should show an error here

              ^^^^^^

lib/ButtonWithoutLinting.dart:15:22: Error: 'should' isn't a type.

          the linter should show an error here

                     ^^^^^^

lib/ButtonWithoutLinting.dart:15:29: Error: Expected ';' after this.

          the linter should show an error here

                            ^^^^

lib/ButtonWithoutLinting.dart:15:34: Error: 'an' isn't a type.

          the linter should show an error here

                                 ^^

lib/ButtonWithoutLinting.dart:15:37: Error: Expected ';' after this.

          the linter should show an error here

                                    ^^^^^

lib/ButtonWithoutLinting.dart:15:43: Error: Expected ';' after this.

          the linter should show an error here

                                          ^^^^

lib/ButtonWithoutLinting.dart:15:43: Error: The getter 'here' isn't defined for the class 'ButtonWithoutLinting'.

 - 'ButtonWithoutLinting' is from 'package:flutter_catc_error_bug/ButtonWithoutLinting.dart' ('lib/ButtonWithoutLinting.dart').

Try correcting the name to the name of an existing getter, or defining a getter or field named 'here'.

          the linter should show an error here

                                          ^^^^





FAILURE: Build failed with an exception.



* Where:

Script 'C:\dev\flutter\packages\flutter_tools\gradle\flutter.gradle' line: 991



* What went wrong:

Execution failed for task ':app:compileFlutterBuildDebug'.

> Process 'command 'C:\dev\flutter\bin\flutter.bat'' finished with non-zero exit value 1



* Try:

Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.



* Get more help at https://help.gradle.org



BUILD FAILED in 14s
Running Gradle task 'assembleDebug'...                             15,4s
Exception: Gradle task assembleDebug failed with exit code 1
```

</details>

<details> 
<summary>flutter doctor -v</summary> 
	
```console
flutter doctor -v
[√] Flutter (Channel stable, 2.0.6, on Microsoft Windows [Version 10.0.19041.450], locale nb-NO)
    • Flutter version 2.0.6 at C:\dev\flutter
    • Framework revision 1d9032c7e1 (2 weeks ago), 2021-04-29 17:37:58 -0700
    • Engine revision 05e680e202
    • Dart version 2.12.3

[√] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at C:\Users\henriks\AppData\Local\Android\Sdk
    • Platform android-30, build-tools 29.0.2
    • ANDROID_HOME = C:\Users\henriks\AppData\Local\Android\Sdk
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b03)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Android Studio (version 3.5)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 42.1.1
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b03)

[√] VS Code (version 1.56.1)
    • VS Code at C:\Users\henriks\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.22.0

[√] Connected device (2 available)
    • Android SDK built for x86 (mobile) • emulator-5554 • android-x86    • Android 10 (API 29) (emulator)
    • Chrome (web)                       • chrome        • web-javascript • Google Chrome 90.0.4430.212

• No issues found!
```

</details>

**Comment by lrhn** [comment]
Reproduces easily. Try this in dartpad:
```dart
catchError(o) { return o; }
class C { } // No `catchError` method needed.
main() {
  C.  // Comment this line out, then lints and errors reappear.
  catchError("x"); // Should be error, C has no such member.

  42.floo(); // Just plain error.
}
```
There are no analyzer warnings or lints. If you add `//` before the `C.`, then multiple lints and errors appear.

Could sound like a bug in special handling of `Future.catchError` which somehow takes down the entire analyzer, and if you run the analyzer on the command line on the same code as above, you get:
```
> dartanalyzer aner.dart
Analyzing aner.dart...
Unhandled exception:
NoSuchMethodError: The getter 'isDartAsyncFuture' was called on null.
Receiver: null
Tried calling: isDartAsyncFuture
#0      Object.noSuchMethod (dart:core-patch/object_patch.dart:54:5)
#1      CatchErrorVerifier.verifyMethodInvocation (package:analyzer/src/error/catch_error_verifier.dart:40:27)
#2      BestPracticesVerifier.visitMethodInvocation (package:analyzer/src/error/best_practices_verifier.dart:596:25)
#3      MethodInvocationImpl.accept (package:analyzer/src/dart/ast/ast.dart:7186:49)
```
(Sounds like a job for null safety! 😉 )

**Comment by scheglov** [comment]
@srawlins

**Comment by srawlins** [comment]
I cannot reproduce. :(

@lrhn's example in the command line, with a recent Dart (`Dart SDK version: 2.14.0-edge.df5d0a5f4e321a6b0ffaa58815e8694f44c8c583 (be) (Mon May 17 21:55:02 2021 +0000) on "macos_x64"`) just shows the expected errors:

```
  error • The method 'catchError' isn't defined for the type 'C'. • a.dart:5:3 • undefined_method
  error • The method 'floo' isn't defined for the type 'int'. • a.dart:7:6 • undefined_method
```

(Works for both `dartanalyzer` and `dart analyze`)

**Comment by scheglov** [comment]
It looks that we have fixed that case of a method invocation without a target.

**Comment by lrhn** [comment]
I was testing on 2.12.4 (latest stable until today) and DartPad. It does seem to be fixed in 2.13.0. Huzzah!
